Regression in 2.24.0: TypeError: Cannot read property 'name' of undefined

**What code were you trying to parse?**

This seems to be an issue with barrel files, and was likely introduced by https://github.com/typescript-eslint/typescript-eslint/commit/133f622f38a286eac45288a9789d2ee529d5e582#diff-68b2b0b0d2d0ff1846b1114472008332R1606

```ts
export * from './app/foo';
```

**What did you expect to happen?**

No error.

**What actually happened?**

```
myrepo\src\public_api.ts
  0:0  error  Parsing error: Cannot read property 'name' of undefined
```

Running with `--debug` gives the following (clipped for brevity):

```
 eslint:file-enumerator Yield: public_api.ts +51ms
  eslint:cascading-config-array-factory Load config files for myrepo\src. +52ms
  eslint:cascading-config-array-factory Cache hit: myrepo\src. +6ms
  eslint:cli-engine Lint myrepo\src\public_api.ts +50ms
  eslint:linter Linting code for myrepo\src\public_api.ts (pass 1) +26ms
  eslint:linter Verify +8ms
  eslint:linter With ConfigArray: myrepo\src\public_api.ts +2ms
  eslint:linter Parsing error: Cannot read property 'name' of undefined
  eslint:linter TypeError: Cannot read property 'name' of undefined
    at Converter.convertNode (myrepo\node_modules\@typescript-eslint\typescript-estree\dist\convert.js:1247:67)
    at Converter.converter (myrepo\node_modules\@typescript-eslint\typescript-estree\dist\convert.js:72:29)
    at Converter.convertChild (myrepo\node_modules\@typescript-eslint\typescript-estree\dist\convert.js:147:21)
    at myrepo\node_modules\@typescript-eslint\typescript-estree\dist\convert.js:203:32
    at Array.map (<anonymous>)
    at Converter.convertBodyExpressions (myrepo\node_modules\@typescript-eslint\typescript-estree\dist\convert.js:202:14)
    at Converter.convertNode (myrepo\node_modules\@typescript-eslint\typescript-estree\dist\convert.js:441:32)
    at Converter.converter (myrepo\node_modules\@typescript-eslint\typescript-estree\dist\convert.js:72:29)
    at Converter.convertProgram (myrepo\node_modules\@typescript-eslint\typescript-estree\dist\convert.js:47:21)
    at Object.astConverter (myrepo\node_modules\@typescript-eslint\typescript-estree\dist\ast-converter.js:26:29) +9ms
```

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `2.24.0` |
| `TypeScript`                           | `3.7.5` |
| `node`                                 | `13.9.0` |
| `npm`                                  | `yarn 1.22.1` |

--- a/packages/typescript-estree/src/convert.ts
+++ b/packages/typescript-estree/src/convert.ts
@@ -1602,7 +1602,12 @@ export class Converter {
             source: this.convertChild(node.moduleSpecifier),
             exportKind: node.isTypeOnly ? 'type' : 'value',
             exported:
-              node.exportClause?.kind === SyntaxKind.NamespaceExport
+              // note - for compat with 3.7.x, where node.exportClause is always undefined and
+              //        SyntaxKind.NamespaceExport does not exist yet (i.e. is undefined), this
+              //        cannot be shortened to an optional chain, or else you end up with
+              //        undefined === undefined, and the true path will hard error at runtime
+              node.exportClause &&
+              node.exportClause.kind === SyntaxKind.NamespaceExport
                 ? this.convertChild(node.exportClause.name)
                 : null,
           });

